bigtable/client/table_bulk_apply_test.cc

```diff
@@ -0,0 +1,377 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/client/chrono_literals.h"
+#include "bigtable/client/data.h"
+
+#include <absl/memory/memory.h>
+#include <gmock/gmock.h>
+#include <google/bigtable/v2/bigtable_mock.grpc.pb.h>
+
+/// Define types and functions used in the tests.
+// TODO(#67) - refactor Mock classes to a testing/ subdirectory.
+namespace {
+class MockClient : public bigtable::ClientInterface {
+ public:
+  MockClient() : mock_stub(new google::bigtable::v2::MockBigtableStub) {}
+
+  MOCK_METHOD1(Open,
+               std::unique_ptr<bigtable::Table>(std::string const &table_id));
+  MOCK_CONST_METHOD0(Stub, google::bigtable::v2::Bigtable::StubInterface &());
+
+  std::unique_ptr<google::bigtable::v2::MockBigtableStub> mock_stub;
+};
+
+class MockReader : public grpc::ClientReaderInterface<
+                       ::google::bigtable::v2::MutateRowsResponse> {
+ public:
+  MOCK_METHOD0(WaitForInitialMetadata, void());
+  MOCK_METHOD0(Finish, grpc::Status());
+  MOCK_METHOD1(NextMessageSize, bool(std::uint32_t *));
+  MOCK_METHOD1(Read, bool(::google::bigtable::v2::MutateRowsResponse *));
```

Comment:
Why is the `google` namespace specified as `::google` while all others are `grpc`, `std`, etc.?

Reply:
You know, thinking about it, this is probably a habit from reading generated proto files, and from the recommendations in the googlemock CookBook.  I do not think I have a good answer (yet) as to what the right approach should be.  Typing `::std` would be extremely odd.  But `testing` is such a common namespace that `::testing` is probably a good defensive move.

I am going to stick with `std::`, `grpc::`, `::google::bigtable`, `::bigtable`, and `::testing` for the time being.  But we need to make a ruling.